crates/matrix-sdk/src/event_cache/room/mod.rs

```diff
@@ -436,14 +436,24 @@ impl RoomEventCacheInner {
                             "not storing previous batch token, because we deduplicated all new sync events"
                         );
 
-                        // Remove the gap we just inserted.
-                        let prev_gap_id = room_events
-                            .rchunks()
-                            .find_map(|c| c.is_gap().then_some(c.identifier()))
-                            .expect("we just inserted the gap beforehand");
-                        room_events
-                            .replace_gap_at([], prev_gap_id)
-                            .expect("we obtained the valid position beforehand");
+                        if let Some(prev_token) = &prev_batch {
+                            // Note: there can't be any race with another task touching the linked
+                            // chunk at this point, because we're using `with_events_mut` which
+                            // guards access to the data.
+                            trace!("removing gap we just inserted");
+
+                            let prev_gap_id = room_events
+                                .rchunks()
+                                .find_map(|c| {
+                                    let gap = as_variant::as_variant!(c.content(), ChunkContent::Gap)?;
+                                    (gap.prev_token == *prev_token).then_some(c.identifier())
```

Comment:
Since we're touching this code, I also made the check more precise: instead of "give me the previous gap, whatever it is", this is now "give me the previous gap with the matching prev-batch token".

Reply:
I would prefer a named function for the code inside `find_map` because it took me quite a while to figure out what this was doing, but I won't insist on it, and I'm happy for the change to come later if you like the idea.

Reply:
Thanks! added a small code comment above this line, expliciting what it does :+1: